### ISSUE TITLE
discv5 topic discovery: efficiency vs resilience tradeoff

### ISSUE THREAD

**Issue by kaiserd** [ethereum/devp2p#199]
[Discv5 topic discovery](https://github.com/ethereum/devp2p/blob/master/discv5/discv5-theory.md) uses `topic radius` as a compromise between efficiency and resilience against eclipse attacks.
I was wondering if topic providers could gain both query efficiency and resilience by placing advertisements at several nodes in ranges of increasing size.
Maybe it would be sufficient to do this without estimating an optimal radius.
Topic providers could start these advertisements in parallel.

Further, topic providers could additionally advertise respective topics as extra fields in their ENRs. This corresponds to advertising topics at radius $2^256$ basically for free, adding additional resilience.

**Comment by fjl** [comment]
Hi,

the topic system is still in the design phase, and a lot of research has happened since the text that's in the spec now was written. In our current research prototype, the concept of 'topic radius' has been removed. I'm in the last stages of completing the prototype and will update the spec soon.

If you are in interested, you can check out our research at

- https://github.com/harnen/service-discovery-paper
- https://github.com/datahop/p2p-service-discovery

**Comment by kaiserd** [comment]
Thank you very much :)! I will read the documents you provided and come back.

**Comment by kaiserd** [comment]
The topic tables proposed in the paper are very similar to what I suggested. So, the problem I described is already covered :). Thank you :).
[Waku v2](https://wakunetwork.com/) could be mentioned as one of the applications benefiting from the enhanced discv5 proposed in the paper; this could further strengthen the motivation.
I will write a PR with things that came to my mind while reading the paper (and a few typo fixes etc); maybe this is helpful.

One more question regarding discovery efficiency:
What is your suggestion for integrating resource restricted devices (e.g. Browser apps, mobile devices) into the discovery process?
Do you envision them being part of discv5 maintaining a routing table or rather just directly query more stable nodes participating in discv5?
(We are planning to specify a `peer-exchange` protocol that allows restricted devices to retrieve random node sets from nodes actively participating in discv5. Resource restricted devices learn about nodes actively participating in discv5 via DNS discovery.
As opposed to directly using nodes learned via DNS as entry points to the target protocol (e.g. Waku relay), asking them for random node sets which are in turn retrieved via discv5 helps load balancing.)

**Comment by kaiserd** [comment]
Yes. Thank you. The issue is covered by the new research :).

Regarding browsers, I was thinking about [portal](https://github.com/ethereum/portal-network-specs/issues/71), and [this](https://github.com/acolytec3/ultralight-browser-client)  discv5 browser implementation.

Yes, I did some NAT related stuff some time ago :). I will have a look and put this as an open issue in our roadmap.
(most of the things (not NAT related) I worked on are on internal Univertiy gitlab instances though 😅).

I opened a [PR](https://github.com/harnen/service-discovery-paper/pull/5) with some comments on the paper's git repo.
As I said there, I could add an estimation of the cost (in expected overlay hops) of random-walk discovery in relation to topic and network size.
We would like to use the new topic discovery for Waku v2 in the future :).
